File: b_rank_new_level_up_problems/b_rank_new_level_up_problems__a_cup_of_water.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;

internal class Program
{
    static void Main()
    {
        int o = 0;
        int[] nx = ReadLine().Split().Select( int.Parse ).ToArray();
        int[] w = new int[nx[0]];
        int[] p = new int[nx[1] + 1];

        foreach( int i in Range( 0, nx[0] ) ) {
            w[i] = int.Parse( ReadLine() );
        }
        p[0] = 1;
        foreach( int i in Range( 0, nx[0] ) ) {
            foreach( int j in Range( 0, nx[1] + 1 ).Reverse() ) {
                if( 0 <= j - w[i] ) {
                    p[j] = System.Math.Max( p[j], p[j - w[i]] );
                }
            }
        }
        foreach( var i in Range( 0, nx[1] + 1 ).Where( i => p[i] == 1 ) ) {
            o = i;
        }
        WriteLine( o );
        return;
    }
}

File: b_rank_new_level_up_problems/b_rank_new_level_up_problems__a_cup_of_water.py
# 実行時間 30ms
import itertools

o = 0
n, x = list(map(int, input().split()))
w = []

for _ in range(n):
    w.append(int(input()))
for i in range(n):
    t = 0
    l = list(itertools.combinations(w, i + 1))

    for j in range(len(l)):
        t = sum(l[j])
        if t <= x and o < t:
            o = t
print(o)
